libs/simdvec/src/main21/java/org/elasticsearch/simdvec/internal/vectorization/MemorySegmentESNextOSQVectorsScorer.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+package org.elasticsearch.simdvec.internal.vectorization;
+
+import jdk.incubator.vector.ByteVector;
+import jdk.incubator.vector.FloatVector;
+import jdk.incubator.vector.IntVector;
+import jdk.incubator.vector.LongVector;
+import jdk.incubator.vector.ShortVector;
+import jdk.incubator.vector.VectorSpecies;
+
+import org.apache.lucene.index.VectorSimilarityFunction;
+import org.apache.lucene.store.IndexInput;
+import org.elasticsearch.simdvec.ESNextOSQVectorsScorer;
+
+import java.io.IOException;
+import java.lang.foreign.MemorySegment;
+
+/** Panamized scorer for quantized vectors stored as a {@link MemorySegment}. */
+public final class MemorySegmentESNextOSQVectorsScorer extends ESNextOSQVectorsScorer {
+
+    private static final VectorSpecies<Integer> INT_SPECIES_128 = IntVector.SPECIES_128;
+
+    private static final VectorSpecies<Long> LONG_SPECIES_128 = LongVector.SPECIES_128;
+    private static final VectorSpecies<Long> LONG_SPECIES_256 = LongVector.SPECIES_256;
+
+    private static final VectorSpecies<Byte> BYTE_SPECIES_128 = ByteVector.SPECIES_128;
+    private static final VectorSpecies<Byte> BYTE_SPECIES_256 = ByteVector.SPECIES_256;
+
+    private static final VectorSpecies<Short> SHORT_SPECIES_128 = ShortVector.SPECIES_128;
+    private static final VectorSpecies<Short> SHORT_SPECIES_256 = ShortVector.SPECIES_256;
+
+    private static final VectorSpecies<Float> FLOAT_SPECIES_128 = FloatVector.SPECIES_128;
+    private static final VectorSpecies<Float> FLOAT_SPECIES_256 = FloatVector.SPECIES_256;
+
+    private final MemorySegment memorySegment;
+    private final MemorySegmentScorer scorer;
+
+    public MemorySegmentESNextOSQVectorsScorer(
+        IndexInput in,
+        byte queryBits,
+        byte indexBits,
+        int dimensions,
+        int dataLength,
+        MemorySegment memorySegment
+    ) {
+        super(in, queryBits, indexBits, dimensions, dataLength);
+        this.memorySegment = memorySegment;
+        if (queryBits == 4 && indexBits == 1) {
```

Comment:
@ChrisHegarty what do you think of this? I am trying to set us up to prevent megamorphic garbage. So, we would end up having this sealed class that is initialized on construction. 

Right now `MemorySegmentESNextOSQVectorsScorer` is pretty much monomorphic, and I didn't want to keep extending it risking performance.